impl/src/main/java/org/jboss/weld/context/beanstore/AttributeBeanStore.java

```diff
@@ -160,7 +188,21 @@ protected NamingScheme getNamingScheme() {
     }
 
     public Iterator<BeanIdentifier> iterator() {
-        return beanStore.iterator();
+        Iterator<BeanIdentifier> iterator;
```

Comment:
> When attaching the store we always copy local entries to the backing store. 

True. However, the local bean store rarely contains some data during attach. Most of the time, it's an empty map. 

> In addition, writes always go directly to the backing store if attached.

True.

> If (!isAttached() || !isAttributeLazyFetchingEnabled()) then return beanStore.iterator();

Ok, let's try this.

> else (meaning both attached and lazy) then it is enough to return getPrefixedAttributeNames()...

I'm not so sure. Is it possible to put a new bean instance in a bean store which is detached?

I like the idea of the on-demand transforming iterator.